### PR TITLE
Move GPU-AV input buffer allocation and initialization to CmdBindDescriptorSets

### PR DESCRIPTION
**Pull request by Tony-LunarG**
Improves GPU-AV performance and memory use running apps with large descriptor sets

### CODE REVIEW

**Review by spencer-lunarg — COMMENTED** [review]
Seems good to move, I like this being moved out of `AllocateValidationResources` to only where it is truly needed

File: layers/gpu_validation.cpp [hunk 1]
```diff
@@ -1915,183 +2109,21 @@ void GpuAssisted::AllocateValidationResources(const VkCommandBuffer cmd_buffer,
...
-                // Pointer to another sets array that points into the bindings array that points into the written array
-                uint32_t *sets_to_bindings = sizes + binding_count;
-                // Pointer to the bindings array that points at the start of the writes in the writes array for each binding
-                uint32_t *bindings_to_written = sets_to_bindings + number_of_sets;
-                // Index of the next entry in the written array to be updated
-                uint32_t written_index = 1 + (number_of_sets * 2) + (binding_count * 2);
-                uint32_t bind_counter = number_of_sets + 1;
-                // Index of the start of the sets_to_bindings array
-                data_ptr[0] = number_of_sets + binding_count + 1;
+        desc_writes[1] = LvlInitStruct<VkWriteDescriptorSet>();
```
[the remark this answers is not in the document]
Comment (Tony-LunarG) [reply layers/gpu_validation.cpp]: Yes, that's better

File: layers/gpu_validation.cpp [hunk 2]
```diff
@@ -2205,10 +2236,15 @@ void gpuav_state::CommandBuffer::Reset() {
...
+    for (auto &buffer_info : per_draw_buffer_list) {
         gpuav->DestroyBuffer(buffer_info);
     }
-    gpuav_buffer_list.clear();
+    per_draw_buffer_list.clear();
+
+    for (auto &buffer_info : di_input_buffer_list) {
+        vmaDestroyBuffer(gpuav->vmaAllocator, buffer_info.buffer, buffer_info.allocation);
+    }
+    di_input_buffer_list.clear();
```
[the remark this answers is not in the document]
Comment (Tony-LunarG) [reply layers/gpu_validation.cpp:2246]: Yes, good catch

File: layers/gpu_validation.cpp [hunk 3]
```diff
@@ -837,9 +837,6 @@ void GpuAssisted::PostCallRecordBindAccelerationStructureMemoryNV(VkDevice devic
 // Free the device memory and descriptor set(s) associated with a command buffer.
 void GpuAssisted::DestroyBuffer(GpuAssistedBufferInfo &buffer_info) {
     vmaDestroyBuffer(vmaAllocator, buffer_info.output_mem_block.buffer, buffer_info.output_mem_block.allocation);
-    if (buffer_info.di_input_mem_block.buffer) {
-        vmaDestroyBuffer(vmaAllocator, buffer_info.di_input_mem_block.buffer, buffer_info.di_input_mem_block.allocation);
-    }
     if (buffer_info.bda_input_mem_block.buffer) {
```
Comment (jeremyg-lunarg) [inline layers/gpu_validation.cpp:840]: I wonder if the buffer device address really should be per-draw allocation still. (not this PR)

**Review by jeremyg-lunarg — APPROVED** [review]
LGTM, glad you got this working!

File: layers/gpu_validation.cpp (hunk 3 above)
Comment (Tony-LunarG) [reply layers/gpu_validation.cpp:840]: Probably not, but so far it hasn't shown up as a bottleneck

**Review by arno-lunarg — APPROVED** [review]

**Review by ncesario-lunarg — APPROVED** [review]

**Pull request merged**